cylc/uiserver/schema.py

```diff
@@ -396,78 +415,299 @@ def run_task_query(conn, workflow):
+            'total_quartiles': [row[28],
+                                row[28] if row[29] is None else row[29],
+                                row[28] if row[30] is None else row[30]],
 
-            'count': row[30]
+            'count': row[31]
         })
 
     return tasks
 
 
-def run_jobs_query(conn, workflow, tasks):
+_JOB_STATUS_TO_STATE = {
+    # task_status: (submit_status, run_status, time_run)
+    TASK_STATUS_SUBMITTED: (0, None, None),
+    TASK_STATUS_SUBMIT_FAILED: (1, None, None),
+    TASK_STATUS_RUNNING: (0, None, True),
+    TASK_STATUS_SUCCEEDED: (0, 0, True),
+    TASK_STATUS_FAILED: (0, 1, True),
+}
+
+
+def _status_to_state(
+    status: str
+) -> Tuple[Optional[int], Optional[int], Optional[bool]]:
+    """Derive job state attributes from job status.
+
+    The time_run cannot be derived from the status so is returned as a boolean.
+
+    Args:
+        status: The job status, e.g. "submitted" or "running".
+
+    Returns:
+        (submit_status, run_status, time_run)
+
+        submit_status:
+            * 0 = successful submission
+            * int = failed submission
+            * None = no submission attempt
+        run_status:
+            * 0 = successful execution
+            * int = failed execution
+            * None = no execution attempt
+        time_run:
+            * True if the job would have started running
+
+    Examples:
+        >>> _status_to_state('running')
+        (0, None, True)
+        >>> _status_to_state('waiting')
+        (None, None, None)
+
+        >>> _state_to_status(*_status_to_state('succeeded'))
+        'succeeded'
+
+    """
+    return _JOB_STATUS_TO_STATE.get(status, (None, None, None))
+
+
+def _state_to_status(
+    submit_status: Optional[int],
+    run_status: Optional[int],
+    time_run: Optional[str],
+) -> str:
+    """Derive job status from state attributes.
+
+    Args:
+        submit_status: Exit code from job submission.
+        run_status: Exit code from job execution.
+        time_run: The time the job reported that it started running.
+            (note this is interpreted as a bool, the exact value is ignored).
+
+    Returns:
+        status: The job status, e.g "submitted" or "running".
+
+    Examples:
+        >>> _state_to_status(0, None, True)
+        'running'
+        >>> _state_to_status(None, None, None)
+        'waiting'
+
+        >>> _status_to_state(_state_to_status(0, 1, True))
+        (0, 1, True)
+
+    """
+    if run_status is not None:
+        if run_status == 0:
+            status = TASK_STATUS_SUCCEEDED
+        else:
+            status = TASK_STATUS_FAILED
+    elif time_run is not None:
+        status = TASK_STATUS_RUNNING
+    elif submit_status is not None:
+        if submit_status == 0:
+            status = TASK_STATUS_SUBMITTED
+        else:
+            status = TASK_STATUS_SUBMIT_FAILED
+    else:
+        status = TASK_STATUS_WAITING
```

Comment:
```suggestion
    for status, state in _JOB_STATUS_TO_STATE.items():
        if state == (submit_status, run_status, time_run):
            return status
    else:
        # I'm not sure about this, but it's a possibility.
        raise ValueError(
            f"Invalid state: {submit_status}, {run_status}, {time_run}"
        )
```
The else is a matter of opinion. But this seems like less data-as-logic.

Reply:
I'm not sure this logic (pre-existing) is 100% the reverse of the dictionary I've added (you can pass improbable combinations of options into this function).

Reply:
No:
* `submit_status` and `run_status` can be `None`, `0` or any positive integer - `>0` is probably the correct test.
* `time_run` is a string(?) or None.

Reply:
Happy to leave pre-existing function alone.

Reply:
> submit_status and can be 0 or any positive integer

No it won't be. These are loaded from the DB and will always be `1 | 0 | None`

Reply:
This is duplicating https://github.com/cylc/cylc-flow/blob/335216488793401ce3f7f90ebf07c11d3187b63c/cylc/flow/data_store_mgr.py#L1655-L1668

can we centralise this logic?

Reply:
> can we centralise this logic?

Sure, but not in a single PR.

Reply:
@oliver-sanders Please can you make a new issue based on this thread.